lib/storage/metricnamestats/tracker.go

```diff
@@ -0,0 +1,266 @@
+type TrackerMetrics struct {
+	CurrentSizeBytes  uint64
+	CurrentItemsCount uint64
+	MaxSizeBytes      uint64
+}
+
+// UpdateMetrics writes internal metrics to the provided object
+func (mt *Tracker) UpdateMetrics(dst *TrackerMetrics) {
+	dst.CurrentSizeBytes = mt.currentSizeBytes.Load()
+	dst.CurrentItemsCount = mt.currentItemsCount.Load()
+	dst.MaxSizeBytes = mt.maxSizeBytes
+}
+
+// IsEmpty checks if internal state has any records
+func (mt *Tracker) IsEmpty() bool {
+	return mt.currentItemsCount.Load() > 0
+}
+
+// Reset cleans stats and saves cache state
+func (mt *Tracker) Reset() {
+	mt.mu.Lock()
+	defer mt.mu.Unlock()
+	mt.store.Clear()
+	mt.currentSizeBytes.Store(0)
+	mt.currentItemsCount.Store(0)
+	mt.creationTs.Store(mt.getCurrentTs())
+	if err := mt.save(); err != nil {
+		logger.Panicf("during Tracker reset cannot save state: %s", err)
+	}
+}
+
+// RegisterIngestRequest tracks metric name ingestion
+func (mt *Tracker) RegisterIngestRequest(metricGroup []byte) {
+	_, ok := mt.store.Load(string(metricGroup))
+	if !ok {
+		if mt.shouldSkipNewItemAdd() {
+			return
+		}
+		var mi statItem
+		mt.store.Store(string(metricGroup), &mi)
+		mt.currentSizeBytes.Add(uint64(len(metricGroup)))
+		mt.currentItemsCount.Add(1)
+	}
+}
+
+// RegisterQueryRequest tracks metric name at query
+//
+// query requests tracking is approximate
+// it may lose newly registered metric during concurrent ingestion
+func (mt *Tracker) RegisterQueryRequest(metricGroup []byte) {
+	v, ok := mt.store.Load(string(metricGroup))
+	if !ok {
+		if mt.shouldSkipNewItemAdd() {
+			return
+		}
+		var mi statItem
+		mi.requestsCount.Store(1)
+		mi.lastRequestTs.Store(mt.getCurrentTs())
+		mt.store.Store(string(metricGroup), &mi)
+		mt.currentSizeBytes.Add(uint64(len(metricGroup)))
+		mt.currentItemsCount.Add(1)
+		return
+	}
+	vi := v.(*statItem)
+	vi.requestsCount.Add(1)
+	vi.lastRequestTs.Store(mt.getCurrentTs())
+}
+
+func (mt *Tracker) shouldSkipNewItemAdd() bool {
+	return mt.currentSizeBytes.Load() > mt.maxSizeBytes
+}
+
+// GetStats returns stats response for the tracked metrics
+func (mt *Tracker) GetStats(limit, lte int, matchPattern string) StatsResult {
+	var result StatsResult
+	result.CollectedSinceTs = mt.creationTs.Load()
+	result.TotalRecords = mt.currentItemsCount.Load()
+	mt.store.Range(func(key, value any) bool {
+		if len(result.Records) >= limit {
+			return false
+		}
+		mi := value.(*statItem)
+		v := mi.requestsCount.Load()
+		if lte < 0 || int(v) <= lte {
+			metricName := key.(string)
+			if len(matchPattern) > 0 && !strings.Contains(metricName, matchPattern) {
+				return true
+			}
+			result.Records = append(result.Records, StatRecord{
+				MetricName:    metricName,
+				RequestCount:  v,
+				LastRequestTs: mi.lastRequestTs.Load(),
+			})
+		}
+		return true
+	})
+	result.sort()
+	return result
+}
+
+// StatsResult defines stats result for GetStats request
+type StatsResult struct {
+	CollectedSinceTs uint64
+	TotalRecords     uint64
```

Comment:
It would be nice to show current utilization of the cache. So user can decide on resetting it

Reply:
Yes, but just returning % should be easier than making promql query from vmui page.